### PR TITLE
[PF-130] RbsDao supports retrieve pool with corresponding number of resources.

### PR DESCRIPTION
**Pull request by yonghaoy**

### CODE REVIEW

File: src/main/java/bio/terra/rbs/db/PoolAndResourceCount.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package bio.terra.rbs.db;
+
+import com.google.auto.value.AutoValue;
+
+/** Represents a {@link Pool} with number of activate resources in the pool. */
+@AutoValue
+abstract class PoolAndResourceCount {
+  abstract Pool pool();
+
+  abstract int resourceCount();
```
Comment (wchamber) [inline src/main/java/bio/terra/rbs/db/PoolAndResourceCount.java:10]: maybe `int activeResources()`? (are we calling them ACTIVE or READY? class comment says "active." consistent terminology is nice again.)

File: src/main/java/bio/terra/rbs/db/RbsDao.java [hunk 2]
```diff
@@ -118,6 +131,19 @@ public void updatePoolsSize(Map<PoolId, Integer> poolsToUpdateSize) {
...
+      (rs, rowNum) ->
+          PoolAndResourceCount.create(
+              Pool.builder()
+                  .id(PoolId.create(rs.getString("id")))
+                  .resourceConfig(deserialize(rs.getString("resource_config")))
+                  .resourceType(ResourceType.valueOf(rs.getString("resource_type")))
+                  .status(PoolStatus.valueOf(rs.getString("status")))
+                  .size(rs.getInt("size"))
+                  .creation(rs.getObject("creation", OffsetDateTime.class).toInstant())
+                  .build(),
```
Comment (wchamber) [inline src/main/java/bio/terra/rbs/db/RbsDao.java:144]: you could use a POOL_ROW_MAPPER here.

File: src/main/java/bio/terra/rbs/db/RbsDao.java [hunk 3]
```diff
@@ -72,6 +72,19 @@ public void createPools(List<Pool> pools) {
     return jdbcTemplate.query(sql, POOL_ROW_MAPPER);
   }
 
+  /** Retrieves all pools and resource count in each pool. */
+  @Transactional(propagation = Propagation.SUPPORTS)
+  public List<PoolAndResourceCount> retrievePoolAndResourceCount() {
+    // TODO: Add filter
+    String sql =
+        "select p.id, p.resource_config, p.resource_type, p.creation, p.size, p.status, "
+            + "(select count (*) from resource where pool_id = p.id) as resource_count "
```
Comment (wchamber) [inline src/main/java/bio/terra/rbs/db/RbsDao.java:81]: this selects all resources counts, but `PoolAndResourceCount` talks about active resources. What do we want?
One thing we can do is have `PoolAndResourceCount` use a `Multiset<ResourceState>` to get counts of all the different resource states. Not sure what's going to be helfpul in the scheduler.

File: src/main/java/bio/terra/rbs/service/resource/FlightScheduler.java [hunk 4]
```diff
@@ -0,0 +1,4 @@
+package bio.terra.rbs.service.resource;
+
+/** Schedulling services periodically scans */
+public class FlightScheduler {}
```
Comment (wchamber) [inline src/main/java/bio/terra/rbs/service/resource/FlightScheduler.java:4]: optional nit: this file doesn't do anything yet.

**Review by wchamber — CHANGES_REQUESTED** [review]

File: src/main/java/bio/terra/rbs/db/RbsDao.java (hunk 2 above)
Comment (yonghaoy) [reply src/main/java/bio/terra/rbs/db/RbsDao.java:144]: Ah, right, thanks

File: src/main/java/bio/terra/rbs/db/RbsDao.java (hunk 3 above)
Comment (yonghaoy) [reply src/main/java/bio/terra/rbs/db/RbsDao.java:81]: I think we only cares about the READY resource state. So add a where clause should be enough?

File: src/main/java/bio/terra/rbs/db/PoolAndResourceCount.java (hunk 1 above)
Comment (yonghaoy) [reply src/main/java/bio/terra/rbs/db/PoolAndResourceCount.java:10]: Changed to activeResourcesCount

File: src/main/java/bio/terra/rbs/db/PoolStatus.java [hunk 5]
```diff
@@ -10,7 +10,8 @@
   /** Active pool, able to handout resources. */
   ACTIVE,
   /**
-   * Deactivated pool, all resources are deactivated or being deactivated, not able to handout resources.
+   * Deactivated pool, all resources are deactivated or being deactivated, not able to handout
+   * resources.
```
Comment (yonghaoy) [inline src/main/java/bio/terra/rbs/db/PoolStatus.java:14]: Modified by spotlessApply

File: src/main/java/bio/terra/rbs/db/RbsDao.java (hunk 3 above)
Comment (wchamber) [reply src/main/java/bio/terra/rbs/db/RbsDao.java]: I'm not sure that's true for the scheduler. We don't want to overschedule resources if there are some resource creations in flight. i.e. we have 10 READY, 2 CREATING, for a pool of size 20. How many flights should we make?
We could make have 2 int fields, resourceReadyCount, resourceCreatingCount, but a `SetMultimap` might scale better.

File: src/main/java/bio/terra/rbs/db/ResourceState.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+ * <p>This is persisted as a string in the database, so the names of the enum values should not be
+ * changed.
+ */
+public enum ResourceState {
+  /** Resource is creating. */
+  CREATING,
+  /** Resource is ready to handout. */
+  READY,
+  /** Resource is handed out. */
+  HANDEDOUT,
```
Comment (wchamber) [inline src/main/java/bio/terra/rbs/db/ResourceState.java:15]: nit: underscored two words, HANDED_OUT?

**Review by wchamber — CHANGES_REQUESTED** [review]
I do think we need to know about the count of CREATING resources for scheduling. Isn't that what we're going to use PoolAndResourceCount for?

File: src/main/java/bio/terra/rbs/db/ResourceState.java (hunk 6 above)
Comment (wchamber) [reply src/main/java/bio/terra/rbs/db/ResourceState.java:15]: also, I don't mind 'handed out' but it is a little awkward to use. any other naming ideas before we commit to the database forever? We could do "USED"? idk, no strong preference.

File: src/main/java/bio/terra/rbs/db/RbsDao.java (hunk 3 above)
Comment (yonghaoy) [reply src/main/java/bio/terra/rbs/db/RbsDao.java]: I think your are right. Updated the sql query and code little bit. PTAL Thanks

File: src/main/java/bio/terra/rbs/db/ResourceState.java (hunk 6 above)
Comment (yonghaoy) [reply src/main/java/bio/terra/rbs/db/ResourceState.java:15]: Okay, changed to used.

File: src/main/java/bio/terra/rbs/db/RbsDao.java [hunk 7]
```diff
@@ -72,6 +74,30 @@ public void createPools(List<Pool> pools) {
     return jdbcTemplate.query(sql, POOL_ROW_MAPPER);
   }
 
+  /** Retrieves all pools and READY resource count in each pool. */
+  @Transactional(propagation = Propagation.SUPPORTS)
+  public List<PoolAndResourceStates> retrievePoolAndResourceStatesCount() {
+    // TODO: Add filter
+    String sql =
+        "select count (*) FILTER (WHERE r.state='READY') as ready_count, "
```
Comment (yonghaoy) [inline src/main/java/bio/terra/rbs/db/RbsDao.java:82]: Not sure if this is the best approach, PTAL, thanks

Comment (wchamber) [reply src/main/java/bio/terra/rbs/db/RbsDao.java:82]: I think it's simpler if we get the counts for all of the resource states, even if we're not examining the handed out state yet. I think a GROUP BY r.state is also more idiomatic than using select subexpressions for each r.state value.
```
select count(*) as resource_count, r.state,
  p.id, p.resource_config, p.resource_type, p.creation, p.size, p.status
from pool p
join resource r on r.pool_id = p.id
group by p.id, r.state;
```
I think that works, but test to verify.
We do need to change the row mapper to be a ResultSetExtractor to group the resource state rows into a single object.
In Janitor, we do this to group the many rows of a TrackedResource with different label key/values into a single object per TrackedResource with all label values. So, not exactly what we're trying to do here, but similar. Let me know if you want to talk more.
https://github.com/DataBiosphere/terra-resource-janitor/blob/e4da842229aacd1e99bfcb1fffb9ac1eadb611dc/src/main/java/bio/terra/janitor/db/JanitorDao.java#L360-L381

File: src/main/java/bio/terra/rbs/db/ResourceState.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+ * <p>This is persisted as a string in the database, so the names of the enum values should not be
+ * changed.
+ */
+public enum ResourceState {
+  /** Resource is creating. */
+  CREATING,
+  /** Resource is ready to handout. */
+  READY,
+  /** Resource is handed out. */
+  USED,
```
Comment (wchamber) [inline src/main/java/bio/terra/rbs/db/ResourceState.java:15]: I really don't know if USED or HANDED_OUT is better. WDYT?

**Review by wchamber — CHANGES_REQUESTED** [review]

File: src/main/java/bio/terra/rbs/db/ResourceState.java (hunk 8 above)
Comment (yonghaoy) [reply src/main/java/bio/terra/rbs/db/ResourceState.java:15]: `USED` sounds wired. I vote for `HANDED_OUT`

File: src/main/java/bio/terra/rbs/db/RbsDao.java [hunk 9]
```diff
@@ -72,6 +77,19 @@ public void createPools(List<Pool> pools) {
     return jdbcTemplate.query(sql, POOL_ROW_MAPPER);
   }
 
+  /** Retrieves all pools and resource count for each state. */
+  @Transactional(propagation = Propagation.SUPPORTS)
+  public List<PoolAndResourceStates> retrievePoolAndResourceStatesCount() {
```
Comment (wchamber) [inline src/main/java/bio/terra/rbs/db/RbsDao.java:82]: nit: `retrievePoolAndResourceStates()`

**Review by wchamber — APPROVED** [review]

**Pull request merged**